src/usr/local/www/firewall_rules_edit.php

```diff
@@ -1088,6 +1084,29 @@ function build_if_list() {
 	return($iflist);
 }
 
+/* Expands missing/blank/"all" icmptypes to a list of specific types. 
+ * Doesn't test $icmptype validity beyond: non-empty string => assumed valid, anything else or 'any'=> any icmptype
+ * If IP protocol undefined in existing/associated rule, then 4th arg can be inet4/inet6/inet46 to force its value, 
+ * or 'unset' to unset the output. Output is placed in $result, passed by reference so we can unset() if required
+*/
+function expand_icmptype($icmptype, $ipprotocol, & $result, $invalid_proto_action = 'unset') {
+	global $icmptypes;
+	// if 1st arg was unset (as could happen historically) then it'll be passed as NULL, so is_string() will suffice
+	if (is_string($icmptype) && strlen($icmptype) > 0 && $icmptype != "any") {
+		// $icmptype appears to contain specific icmptype(s)
+		$result = $icmptype;
+	} elseif (is_string($ipprotocol) && array_key_exists($ipprotocol, $icmptypes)) {
+		// $icmptype has no (or blank) icmptype,or is "all"  meaning "all/any", select all subtypes
+		$result = implode(',', $icmptypes[$ipprotocol]['icmptypes']);
+	} elseif ($invalid_proto_action != 'unset') {
+		// ipprotocol undefined or invalid, force value to whatever we were passed, should be inet4/6/46
+		$result = implode(',', $icmptypes[$invalid_proto_action]['icmptypes']);
+	? else {
```

Comment:
`Parse error: syntax error, unexpected '?' in /usr/local/www/firewall_rules_edit.php on line 1098 `
That `?` should be a `{`